src/ast_pattern.mli

```diff
@@ -210,3 +210,6 @@ type context
 
 val of_func : (context -> Location.t -> 'a -> 'b -> 'c) -> ('a, 'b, 'c) t
 val to_func : ('a, 'b, 'c) t -> context -> Location.t -> 'a -> 'b -> 'c
+
+val fail : Location.t -> string -> _
+(** Call from [of_func]'s argument when the pattern does not match. *)
```

Comment:
Might be worth mentioning the `string` argument should describe the expected AST's shape.